File: wagerproof-ios-native/Wagerproof/Features/Agents/Creation/Inputs/OddsInput.swift
import SwiftUI
import WagerproofDesign

/// Odds floor/ceiling input. Ports `components/agents/inputs/OddsInput.tsx`.
///
/// Behavior:
///   - `favorite` type → valid range -500...-100 (e.g. "-200")
///   - `underdog` type → valid range +100...+500 (e.g. "+150")
///   - Tapping "No limit" chip toggles between an active value and `nil`.
///   - The value is clamped to the valid range on commit (focus loss).
struct OddsInput: View {
    enum InputType { case favorite, underdog }

    @Binding var value: Int?
    let label: String
    let type: InputType

    @State private var text: String = ""
    @State private var validationError: String? = nil
    @FocusState private var isFocused: Bool

    private var isFavorite: Bool { type == .favorite }
    private var minValue: Int { isFavorite ? -500 : 100 }
    private var maxValue: Int { isFavorite ? -100 : 500 }
    private var placeholder: String { isFavorite ? "-200" : "+150" }

    private var helperText: String {
        isFavorite
            ? "Skip heavier favorites once the price gets too steep."
            : "Only allow plus-money dogs that clear your floor."
    }

    private var isNoLimit: Bool { value == nil }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack(alignment: .top, spacing: 12) {
                VStack(alignment: .leading, spacing: 4) {
                    Text(label)
                        .font(.system(size: 16, weight: .bold))
                        .foregroundStyle(Color.appTextPrimary)
                    Text(helperText)
                        .font(.system(size: 12))
                        .foregroundStyle(Color.appTextSecondary)
                        .fixedSize(horizontal: false, vertical: true)
                }
                Spacer(minLength: 8)
                Text(isNoLimit ? "No limit" : formatOdds(value ?? 0))
                    .font(.system(size: 11, weight: .bold))
                    .tracking(0.3)
                    .padding(.horizontal, 10)
                    .padding(.vertical, 6)
                    .background(
                        Capsule()
                            .fill(isNoLimit ? Color.appBorder.opacity(0.5) : Color(hex: 0x3B82F6).opacity(0.22))
                    )
                    .foregroundStyle(isNoLimit ? Color.appTextSecondary : Color(hex: 0xBFDBFE))
            }

            HStack(spacing: 12) {
                TextField(placeholder, text: $text)
                    .keyboardType(.numbersAndPunctuation)
                    .multilineTextAlignment(.center)
                    .font(.system(size: 18, weight: .semibold))
                    .focused($isFocused)
                    .disabled(isNoLimit)
                    .padding(.vertical, 12)
                    .padding(.horizontal, 16)
                    .background(
                        RoundedRectangle(cornerRadius: 12, style: .continuous)
                            .fill(Color.appBorder.opacity(0.3))
                    )
                    .overlay(
                        RoundedRectangle(cornerRadius: 12, style: .continuous)
                            .strokeBorder(validationError != nil ? Color(hex: 0xEF4444) : Color.appBorder, lineWidth: 1)
                    )
                    .opacity(isNoLimit ? 0.5 : 1)
                    .onChange(of: text) { _, newValue in
                        handleTextChange(newValue)
                    }
                    .onChange(of: isFocused) { _, focused in
                        if !focused { handleBlur() }
                    }

                Button {
                    toggleNoLimit()
                } label: {
                    Text("No limit")
                        .font(.system(size: 12, weight: .semibold))
                        .padding(.horizontal, 12)
                        .padding(.vertical, 8)
                        .foregroundStyle(isNoLimit ? Color(hex: 0x00E676) : Color.appTextSecondary)
                        .background(
                            Capsule()
                                .fill(isNoLimit ? Color(hex: 0x00E676).opacity(0.15) : Color.clear)
                        )
                        .overlay(
                            Capsule()
                                .strokeBorder(isNoLimit ? Color(hex: 0x00E676) : Color.appBorder, lineWidth: 1)
                        )
                }
                .buttonStyle(.plain)
            }

            if let validationError {
                Text(validationError)
                    .font(.system(size: 12))
                    .foregroundStyle(Color(hex: 0xEF4444))
            }
        }
        // Outer card chrome removed — renders as a plain Form row.
        // All parsing/validation/clamping logic preserved verbatim above.
        .onAppear { syncText(from: value) }
        .onChange(of: value) { _, newValue in
            // External writes (preset apply, etc.) re-sync the text field.
            if !isFocused { syncText(from: newValue) }
        }
    }

    // MARK: - Helpers

    private func syncText(from v: Int?) {
        text = v.map(formatOdds) ?? ""
    }

    private func formatOdds(_ odds: Int) -> String {
        odds >= 0 ? "+\(odds)" : "\(odds)"
    }

    private func parseOdds(_ raw: String) -> Int? {
        let cleaned = raw.unicodeScalars.filter { CharacterSet(charactersIn: "0123456789-").contains($0) }
        let str = String(String.UnicodeScalarView(cleaned))
        return str.isEmpty ? nil : Int(str)
    }

    private func validate(_ odds: Int?) -> String? {
        guard let odds else { return nil }
        if isFavorite {
            if odds > -100 { return "Favorite odds must be -100 or lower" }
            if odds < -500 { return "Favorite odds cannot be lower than -500" }
        } else {
            if odds < 100 { return "Underdog odds must be +100 or higher" }
            if odds > 500 { return "Underdog odds cannot exceed +500" }
        }
        return nil
    }

    private func handleTextChange(_ newText: String) {
        let parsed = parseOdds(newText)
        validationError = validate(parsed)
        if validationError == nil, let parsed {
            value = parsed
        }
    }

    private func handleBlur() {
        guard !text.isEmpty else { return }
        guard let parsed = parseOdds(text) else {
            syncText(from: value)
            return
        }
        let clamped = max(minValue, min(maxValue, parsed))
        value = clamped
        text = formatOdds(clamped)
        validationError = nil
    }

    private func toggleNoLimit() {
        if value == nil {
            // Hop to a sensible default per direction (mirrors RN).
            let next = isFavorite ? -200 : 150
            value = next
            text = formatOdds(next)
        } else {
            value = nil
            text = ""
        }
        validationError = nil
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/agents/creation/inputs/OddsInput.kt
package com.wagerproof.app.features.agents.creation.inputs

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.core.design.tokens.AppColors

/** Whether the input governs a favorite floor or an underdog floor. */
enum class OddsInputType { FAVORITE, UNDERDOG }

/**
 * American-odds floor/ceiling input for Step 4's "Odds Limits". Port of iOS
 * `OddsInput`.
 *   - FAVORITE → valid -500..-100 (e.g. "-200")
 *   - UNDERDOG → valid +100..+500 (e.g. "+150")
 *   - "No limit" chip toggles between an active value and null.
 *   - Value is clamped to the valid range on blur (focus loss).
 */
@Composable
fun OddsInput(
    value: Int?,
    onValueChange: (Int?) -> Unit,
    label: String,
    type: OddsInputType,
    modifier: Modifier = Modifier,
) {
    val isFavorite = type == OddsInputType.FAVORITE
    val minValue = if (isFavorite) -500 else 100
    val maxValue = if (isFavorite) -100 else 500
    val placeholder = if (isFavorite) "-200" else "+150"
    val helperText = if (isFavorite)
        "Skip heavier favorites once the price gets too steep."
    else
        "Only allow plus-money dogs that clear your floor."

    val isNoLimit = value == null

    fun formatOdds(odds: Int): String = if (odds >= 0) "+$odds" else "$odds"
    fun parseOdds(raw: String): Int? {
        val str = raw.filter { it.isDigit() || it == '-' }
        return if (str.isEmpty() || str == "-") null else str.toIntOrNull()
    }
    fun validate(odds: Int?): String? {
        if (odds == null) return null
        return if (isFavorite) {
            when {
                odds > -100 -> "Favorite odds must be -100 or lower"
                odds < -500 -> "Favorite odds cannot be lower than -500"
                else -> null
            }
        } else {
            when {
                odds < 100 -> "Underdog odds must be +100 or higher"
                odds > 500 -> "Underdog odds cannot exceed +500"
                else -> null
            }
        }
    }

    var text by remember { mutableStateOf(value?.let { formatOdds(it) } ?: "") }
    var validationError by remember { mutableStateOf<String?>(null) }
    var isFocused by remember { mutableStateOf(false) }

    // External writes (preset apply, etc.) re-sync the text when not focused.
    LaunchedEffect(value) {
        if (!isFocused) text = value?.let { formatOdds(it) } ?: ""
    }

    Column(modifier = modifier.fillMaxWidth().padding(vertical = 6.dp)) {
        Row(verticalAlignment = Alignment.Top) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = label,
                    color = AppColors.appTextPrimary,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Bold,
                )
                Text(
                    text = helperText,
                    color = AppColors.appTextSecondary,
                    fontSize = 12.sp,
                )
            }
            Spacer(Modifier.width(12.dp))
            Text(
                text = if (isNoLimit) "No limit" else formatOdds(value ?: 0),
                color = if (isNoLimit) AppColors.appTextSecondary else Color(0xFFBFDBFE),
                fontSize = 11.sp,
                fontWeight = FontWeight.Bold,
                modifier = Modifier
                    .background(
                        color = if (isNoLimit) AppColors.appBorder.copy(alpha = 0.5f)
                        else AppColors.appAccentBlue.copy(alpha = 0.22f),
                        shape = RoundedCornerShape(50),
                    )
                    .padding(horizontal = 10.dp, vertical = 6.dp),
            )
        }

        Spacer(Modifier.width(12.dp))

        Row(verticalAlignment = Alignment.CenterVertically) {
            val borderColor = if (validationError != null) AppColors.appAccentRed else AppColors.appBorder
            Box(
                modifier = Modifier
                    .weight(1f)
                    .background(AppColors.appBorder.copy(alpha = 0.3f), RoundedCornerShape(12.dp))
                    .border(1.dp, borderColor, RoundedCornerShape(12.dp))
                    .padding(vertical = 12.dp, horizontal = 16.dp),
                contentAlignment = Alignment.Center,
            ) {
                if (text.isEmpty()) {
                    Text(
                        text = placeholder,
                        color = AppColors.appTextSecondary.copy(alpha = if (isNoLimit) 0.5f else 1f),
                        fontSize = 18.sp,
                        fontWeight = FontWeight.SemiBold,
                        textAlign = TextAlign.Center,
                    )
                }
                BasicTextField(
                    value = text,
                    onValueChange = { newText ->
                        text = newText
                        val parsed = parseOdds(newText)
                        validationError = validate(parsed)
                        if (validationError == null && parsed != null) onValueChange(parsed)
                    },
                    enabled = !isNoLimit,
                    singleLine = true,
                    textStyle = TextStyle(
                        color = AppColors.appTextPrimary.copy(alpha = if (isNoLimit) 0.5f else 1f),
                        fontSize = 18.sp,
                        fontWeight = FontWeight.SemiBold,
                        textAlign = TextAlign.Center,
                    ),
                    cursorBrush = SolidColor(AppColors.appPrimary),
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                    modifier = Modifier
                        .fillMaxWidth()
                        .onFocusChanged { state ->
                            if (isFocused && !state.isFocused) {
                                // Blur: clamp a parseable value into range, else resync.
                                if (text.isNotEmpty()) {
                                    val parsed = parseOdds(text)
                                    if (parsed == null) {
                                        text = value?.let { formatOdds(it) } ?: ""
                                    } else {
                                        val cl = parsed.coerceIn(minValue, maxValue)
                                        onValueChange(cl)
                                        text = formatOdds(cl)
                                        validationError = null
                                    }
                                }
                            }
                            isFocused = state.isFocused
                        },
                )
            }

            Spacer(Modifier.width(12.dp))

            val noLimitFg = if (isNoLimit) AppColors.brandGreenBright else AppColors.appTextSecondary
            Text(
                text = "No limit",
                color = noLimitFg,
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier
                    .background(
                        color = if (isNoLimit) AppColors.brandGreenBright.copy(alpha = 0.15f) else Color.Transparent,
                        shape = RoundedCornerShape(50),
                    )
                    .border(
                        1.dp,
                        if (isNoLimit) AppColors.brandGreenBright else AppColors.appBorder,
                        RoundedCornerShape(50),
                    )
                    .clickableNoRipple {
                        if (value == null) {
                            val next = if (isFavorite) -200 else 150
                            onValueChange(next)
                            text = formatOdds(next)
                        } else {
                            onValueChange(null)
                            text = ""
                        }
                        validationError = null
                    }
                    .padding(horizontal = 12.dp, vertical = 8.dp),
            )
        }

        validationError?.let { err ->
            Text(
                text = err,
                color = AppColors.appAccentRed,
                fontSize = 12.sp,
                modifier = Modifier.padding(top = 4.dp),
            )
        }
    }
}
